File: Frogger/tree.cpp
#include "tree.h"

Tree::Tree(float x, float y, float z) : StaticObject(x, y, z) {
	
	billboard = new Surface();
	setupObjects();
}

void Tree::setupObjects(){

	vectorOrientation[0] = 1.0f;
	vectorOrientation[1] = 0.0f;
	vectorOrientation[2] = 0.0f;

	float body_amb[] = { 0.0f, 1.0f, 0.0f, 1.0f };
	float body_diff[] = { 0.0f, 1.0f, 0.0f, 1.0f };
	float body_spec[] = { 0.0f, 1.0f, 0.0f, 1.0f };

	//Tree
	billboard->addTexture(0, "assets/tree.tga");
	billboard->setColor(VSResourceLib::AMBIENT, body_amb);
	billboard->setColor(VSResourceLib::DIFFUSE, body_diff);
	billboard->setColor(VSResourceLib::SPECULAR, body_spec);
}

void Tree::render(VSShaderLib shader){
	
	vsml->loadIdentity(VSMathLib::MODEL);
	vsml->pushMatrix(VSMathLib::MODEL);

	vsml->translate(position[0], position[1], position[2]);
	
	if ((angle < 0.99990) && (angle > -0.9999))
		vsml->rotate(acos(-angle) * 180.0f / 3.14f, auxVector[0], auxVector[1], auxVector[2]);

	vsml->scale(1.0f, 10.0f, 8.0f);
	billboard->render(shader);
	
	vsml->popMatrix(VSMathLib::MODEL);
}

void Tree::updateRotation(float* frogPos){
	
	vector[0] = frogPos[0] - position[0];
	vector[1] = 0.0f;
	vector[2] = frogPos[2] - position[2];

	vsml->normalize(vector);
	vsml->crossProduct(vector, vectorOrientation, auxVector);
	angle = vsml->dotProduct(vector, vectorOrientation);
}

File: FroggerJS/tree.js
function Tree(x, y, z) {

	this.x = x;
	this.y = y;
	this.z = z;

	this.angle = 0.0;
	this.vectorOrientation = [];
	this.auxVector = [];
	this.vector = [];

	this.surface = new Surface();
	this.setupObjects();
}

Tree.prototype.setupObjects = function() {

	this.vectorOrientation[0] = 1.0;
	this.vectorOrientation[1] = 0.0;
	this.vectorOrientation[2] = 0.0;

	this.surface.setColor("ambient", [0.0, 1.0, 0.0, 1.0]);
	this.surface.setColor("diffuse", [0.0, 1.0, 0.0, 1.0]);
	this.surface.setColor("ambient", [0.0, 1.0, 0.0, 1.0]);
	this.surface.setColor("emissive", [0.0, 0.0, 0.0, 1.0]);
	this.surface.setColor("shininess", [0.0, 0.0, 0.0, 1.0]);
}

Tree.prototype.render = function() {

	mat4.identity(mMatrix);

	pushMatrix();

	mat4.translate(mMatrix, [this.x, this.y, this.z]);

	if ((this.angle < 0.99990) && (this.angle > -0.9999)) {
		//GLMatrix 1.3.7 rotate function doesn't work as expected. We need to upgrade lib version in the future.
		mat4.rotate(mMatrix, Math.acos(-this.angle) * (180.0 / 3.14), this.auxVector);
	}

	mat4.scale(mMatrix, [1.0, 10.0, 8.0]);
	this.surface.render();

	popMatrix();
}

Tree.prototype.updateRotation = function() {

	this.vector[0] = frog.x - this.x;
	this.vector[1] = 0.0;
	this.vector[2] = frog.z - this.z;

	vec3.normalize(this.vector);
	vec3.cross(this.vector, this.vectorOrientation, this.auxVector);
	this.angle = vec3.dot(this.vector, this.vectorOrientation);
}